vnext/local-cli/runWindows/utils/autolink.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ * @format
+ */
+// @ts-check
+
+const execSync = require('child_process').execSync;
+const path = require('path');
+const fs = require('fs');
+
+function updateAutoLink(verbose) {
+  const execString =
+    'react-native config';
+  let output;
+  try {
+    console.log('Running react-natice config...');
+    output = execSync(execString).toString();
+    if (verbose)
+    {
+      console.log(output);
+    }
+
+    const config = JSON.parse(output);
+    const windowsPlatformConfig = config['project']['windows'];
+    const cppProjFile= windowsPlatformConfig?windowsPlatformConfig.cppProjFile:null;
+    if (cppProjFile == null){
+      console.log('AutoLink currently is only supported on cpp/winrt main project.');
+      return;
+    }
+    console.log('cppProjFile:' + cppProjFile);
+
+    const sourceDir = config['project']['windows'].sourceDir;
+    console.log('sourceDir:' + sourceDir);
+
+    const projectName = config['project']['windows'].projectName;
+    console.log('projectName:' + projectName);
+    
+    //#1. update nativeModules.g.h
+    const generatedHeader =  path.join(sourceDir, projectName, 'nativeModules.g.h');
+    if (!fs.existsSync(generatedHeader))
+    {
+      console.log('autoLink can not locate generated header file:' + generatedHeader);
+      return;
+    }
+
+    var generatedIncludes = '// NativeModules.g.h -- contents generated by "react-native run-windows"';
+    const dependencies = config['dependencies'];
+    var packageRegistrations = '#define RegisterNativeModulePackages';
```

Comment:
>#define RegisterNativeModulePackages [](start = 32, length = 36)

Actually, yes, it must be possible to generate an inline function here.
It will be have an empty body if nothing to register.

While macros are still used in C++ as necessary 'evil' to accomplish some work, it is better to choose more C++ish solution when it is possible.

Reply:
I still like macro better, for code-gen simplicity. Otherwise, I need to pass the packageProviders parameter and a bunch of additional include and namespae statements.

---
In reply to: [391162936](https://github.com/microsoft/react-native-windows/pull/4293#discussion_r391162936) [](ancestors = 391162936)